staticaddr/loopin/actions.go

```diff
@@ -0,0 +1,1057 @@
+		cancelInvoiceSubscription()
+	}
+
+	cancelInvoice := func() {
+		f.Errorf("timeout waiting for invoice to be " +
+			"paid, canceling invoice")
+
+		// Cancel the lndclient invoice subscription.
+		cancelInvoiceSubscription()
+
+		err = f.cfg.InvoicesClient.CancelInvoice(ctx, f.loopIn.SwapHash)
+		if err != nil {
+			f.Warnf("unable to cancel invoice "+
+				"for swap hash: %v", err)
+		}
+	}
+
+	for {
+		select {
+		case <-htlcConfChan:
+			f.Infof("htlc tx confirmed")
+
+			htlcConfirmed = true
+
+		case err = <-htlcErrConfChan:
+			f.Errorf("htlc tx conf chan error: %v", err)
+
+		case <-reorgChan:
+			// A reorg happened. We invalidate a previous htlc
+			// confirmation and re-register for the next
+			// confirmation.
+			htlcConfirmed = false
+
+			htlcConfChan, htlcErrConfChan, err = registerHtlcConf()
+			if err != nil {
+				f.Errorf("unable to monitor htlc tx "+
+					"confirmation: %v", err)
+			}
+
+		case <-deadlineChan:
+			// If the server didn't pay the invoice on time, we
+			// cancel the invoice and keep monitoring the htlc tx
+			// confirmation. We also need to unlock the deposits to
+			// re-enable them for loop-ins and withdrawals.
+			cancelInvoice()
+
+			event := f.UnlockDepositsAction(ctx, nil)
+			if event != fsm.OnError {
+				f.Errorf("unable to unlock deposits after " +
+					"payment deadline")
+			}
+
+		case currentHeight := <-blockChan:
+			// If the htlc is confirmed but blockChan fires before
+			// htlcConfChan, we would wrongfully assume that the
+			// htlc tx was not confirmed which would lead to
+			// returning OnSwapTimedOut in the code below. This in
+			// turn would prevent us from sweeping the htlc timeout
+			// path back to us.
+			// Hence, we delay the timeout check here by one block
+			// to ensure that htlcConfChan fires first.
+			if !f.loopIn.isHtlcTimedOut(currentHeight - 1) {
+				// If the htlc hasn't timed out yet, we continue
+				// monitoring the htlc confirmation and the
+				// invoice settlement.
+				continue
+			}
+
+			f.Infof("htlc timed out at block height %v",
+				currentHeight)
+
+			// If the timeout path opened up we consider the swap
+			// failed and cancel the invoice.
+			cancelInvoice()
+
+			if !htlcConfirmed {
+				f.Infof("swap timed out, htlc not confirmed")
+
+				// If the htlc hasn't confirmed but the timeout
+				// path opened up, and we didn't receive the
+				// swap payment, we consider the swap attempt to
+				// be failed. We cancelled the invoice, but
+				// don't need to unlock the deposits because
+				// that happened when the payment deadline was
+				// reached.
+				return OnSwapTimedOut
+			}
+
+			// If the htlc has confirmed and the timeout path has
+			// opened up we sweep the funds back to us.
+			err = f.cfg.DepositManager.TransitionDeposits(
+				ctx, f.loopIn.Deposits,
+				deposit.OnSweepingHtlcTimeout,
+				deposit.SweepHtlcTimeout,
+			)
+			if err != nil {
+				log.Errorf("unable to transition "+
```

Comment:
In this case manual intervention would be needed iiuc?

Reply:
The timeout transaction would be published and funds would be safe, but the client would show the deposits in state `LoopingIn`. So for sanity the state would have to be changed manually, yes.